tf_helpers/lax.py

```diff
@@ -0,0 +1,505 @@
+  """ The general dot operation for TensorFlow.
+
+  An equivalent general dot operation as that in JAX -
+     <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+  Although there is an implementation in TF XLA, avoid directly using XLA when
+  possible.
+
+  e.g., non-batched: ij,jk->ik
+        batched: ijk,ikl->ijl
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    dimension_numbers: (Tuple[Tuple[Sequence[int], Sequence[int]],
+      Tuple[Sequence[int], Sequence[int]]]) – a tuple of tuples of the form
+      ((lhs_contracting_dims, rhs_contracting_dims), (lhs_batch_dims, rhs_batch_dims))
+
+  Returns:
+    An array that contains the result.
+  """
+  char_list = list(string.ascii_lowercase)
+  char_list = char_list[8:] + char_list[:8]
+  lhs_rank, rhs_rank = len(lhs.shape), len(rhs.shape)
+  lhs_rep = char_list[:lhs_rank]
+  rhs_rep = char_list[lhs_rank:lhs_rank+rhs_rank]
+  contraction, batch = dimension_numbers
+  lhs_contraction, rhs_contraction = contraction
+  if len(lhs_contraction) != len(rhs_contraction):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of contraction dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_contraction), len(rhs_contraction)))
+  lhs_batch, rhs_batch = batch
+  if len(lhs_batch) != len(rhs_batch):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of batch dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_batch), len(rhs_batch)))
+
+  if len(lhs_batch) == 0 and len(rhs_batch) == 0:
+    return _non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction)
+
+  if (lhs_rank == rhs_rank == 3 and lhs_batch == (0,) and rhs_batch == (0,)
+      and lhs_contraction == (2,) and rhs_contraction == (1,)):
+    return np.asarray(tf.linalg.matmul(lhs, rhs))
+
+  for i in range(len(lhs_contraction)):
+    rhs_rep[rhs_contraction[i]] = lhs_rep[lhs_contraction[i]]
+  for i in range(len(lhs_batch)):
+    rhs_rep[rhs_batch[i]] = lhs_rep[lhs_batch[i]]
+
+  output_rep = _compose_output_rep(lhs_rep, rhs_rep, lhs_contraction,
+                                  rhs_contraction, lhs_batch, rhs_batch)
+  equation = ''.join(lhs_rep) + ',' + ''.join(rhs_rep) + "->" + output_rep
+  return np.asarray(tf.einsum(equation, lhs, rhs))
+
+
+def reduce_window(inputs, init_value, reducer, window_dimensions, strides,
+                  padding, base_dilation=None, window_dilation=None):
+  if reducer not in [np.max, np.add]:
+    raise TypeError("Only max pooling and average/sum pooling are supported.")
+
+  # Note that there is no need to send in the parameter data format since the
+  # input is already of default data format - "N...C". The adjustments of the
+  # input shape is already finished in apply_fun of Pooling in stax.
```

Comment:
Is this referring to your upcoming modification of `stax.py` pooling layers? If so, I believe it may be cleaner to have the adjustments be done here, to match the `lax.reduce_window` API precisely, and not have to prepare the inputs to each function call. (IIUC, you'd need to reshape `inputs` to `(1, 1) + inputs.shape` (https://www.tensorflow.org/api_docs/python/tf/nn/pool). But lmk if I'm missing something here!

Reply:
Yes, Roman, you pointed out the right direction! But things are a bit complex here and it really depends on cases. The above that you mentioned fits the case where the inputs do not contain batch and channel dimensions, but there are also cases where only batch dimension is there, the window dimensions contain unnecessary batch and channel dimensions where we need to remove them, etc., if I remember correctly. So I guess it would be more flexible to keep it in the current form?

Reply:
Hi, Roman, thanks for paying attention to the details here! Below is some elaboration, please correct me if I am wrong!

In terms of point one you mentioned, `N...C` is the default format if you look at the parameter requirements in the link that you provided. The original texts on official documents are: 
```
A string or None. Specifies whether the channel dimension of the input and output is the last dimension (default, or if data_format does not start with "NC"), or the second dimension (if data_format starts with "NC"). For N=1, the valid values are "NWC" (default) and "NCW". For N=2, the valid values are "NHWC" (default) and "NCHW". For N=3, the valid values are "NDHWC" (default) and "NCDHW".
```

In terms of the second point that you mentioned, there is key design difference between TensorFlow `pool` and JAX `reduce_window`: In TF `pool`, <b>the window shapes only require spatial dimensions</b>, so no matter whether there are batch and channel dimensions, no matter how many batch and channel dimensions are there, the filters are assumed to be the same across batches by default. So for example, if the input matrix is of form `N...C` of rank `n+2`, then the window shape is required to be of rank `n`. <b>However, in JAX, the window shapes are of same rank as the input matrix, and this is much more powerful because it allows much more possibilities in terms of batch and channel dimensions</b>. In TensorFlow, if I want to build `reduce_window` on top of `pool`, it is going to be super hard to recognize which specific situation it is, unless some more parameters are sent in. 

Again, please correct me if I am wrong.

Reply:
Besides, in TensorFlow, there is an underlying checker that checks `rank(window_shape) + 2 == rank(input)`, so it is like hard requirement.

Reply:
Solved in https://github.com/google/neural-tangents/pull/59/commits/9357a243ec8d3c685cee47faee2bd6c94dc3a0fa; Thanks Roman for bringing this up.

Reply:
@romanngg Hi Roman, I am testing the new TF `reduce_window`, and sadly I find a flaw that is due to TF `pool` limitation. TF `pool` currently only supports 1 <= spatial dimensions <=3, and this is going to be very limited for our inputs...which means we can only have 3 dimensional inputs (including batch and channel), not going to handle the most common 4 dimensional cases.

Reply:
Ah, that's a bummer! Did you find a solution to it / perhaps the TF team could suggest something here? (maybe we could call the XLA method here?)

In NT pooling over 4 spatial dimensions is a common operation if average pooling layers are used in the network (https://github.com/google/neural-tangents/blob/8b79dbeb32e9df7a6323df43ede898b6897eea01/neural_tangents/stax.py#L1291), and I guess we also would have the issue of breaking the lax correspondence which I imagine will cause us a fair amount of headache. Further, in general, NT currently is dimensionality-agnostic, so this would also break the TF interface at inputs having more than 4 spatial dimensions.

Workarounds that I see are:
1) implement ND-pooling as a sequence of lower-dimensional poolings, like separable convolutions;
2) Implement ND-pooling as a convolution, if by any chance it's supported in higher dimensions in TF.
But ofc both would be slower and not particularly nice, so I'm not sure here. Lmk what you think!

Reply:
Hi Roman, thanks for the thoughts here! TF convolution also only supports 1 to 3 number of spatial dimensions. I will send an email and let's see what my mentors at TF team say about this.